Kaptain 2.0.0 Docs configuration folder (#4311)
mesosphere/dcos-docs-site: pages/dkp/kaptain/2.0.0/configuration/auto-cleanup-settings/index.md

The Notebook Controller provides functionality called “notebook culling”  which can scale down idle notebooks. Scaling down notebooks frees up the resources allocated to those notebooks and makes them available to other workloads. 

Jupyter Notebook exposes an endpoint reporting the last activity within a notebook; if the last activity exceeds the configured limit, the controller scales the underlying StatefulSet to 0 replicas.

When a notebook is up and running, it is displayed as active in the UI and has one replica in the StatefulSet:

![notebook-running.png](../../img/notebook-running.png)

```bash
kubectl get notebooks.kubeflow.org && kubectl get statefulsets

NAME       AGE
notebook   9m36s
NAME       READY   AGE
notebook   1/1     9m37s
```

After the notebook has idled longer than the specified culling time, it is scaled down:

![notebook-stopped.png](../../img/notebook-stopped.png)

```bash
kubectl get notebooks.kubeflow.org && kubectl get statefulsets

NAME       AGE
notebook   12m
NAME       READY   AGE
notebook   0/0     13m
```

> You can *resume* the notebook from the UI later, and the corresponding StatefulSet is scaled back to 1 replica if there are sufficient resources on the cluster.   The workspace volume is automatically attached to the resumed notebook.

Author: ```suggestion
You can *resume* the notebook from the UI later, and the corresponding StatefulSet is scaled back to 1 replica if there are sufficient resources on the cluster. The workspace volume is automatically attached to the resumed notebook.
```

dsl-control-structures-rugqkrh-1-2276733026-3461867059   0/2     Completed   0          3m51s
dsl-control-structures-rugqkrh-1-2276733026-4042755208   0/2     Completed   0          4m1s
```

### Using Python DSL for setting Pipeline TTL 
Kubeflow Pipelines provide a Python Domain Specific Language (DSL) that allows you to specify a time-to-live (TTL) for the submitted Pipeline. Here is an excerpt from the [Pipeline tutorial][pipeline-tutorial]:
```python
@dsl.pipeline(
    name="End-to-End MNIST Pipeline",
    description="A sample pipeline to demonstrate multi-step model training, evaluation, export, and serving",
)
def mnist_pipeline(
    input_bucket: str = "tutorial",
    ...
):
    train_and_serve(
        input_bucket=input_bucket,
        ...
    )
    ...
    # TTL for the workflow to persist after completion (1 hour)
    dsl.get_pipeline_conf().set_ttl_seconds_after_finished(60 * 60)
```

This setting specifies the `ttlSecondsAfterFinished` property in the Argo Workflow definition specifying the amount of time the workflow will persist before it is cleaned up by the dedicated controller. 

### Setting global TTL for completed Pipelines

> Kaptain has a global configuration property that allows you to set the default TTL for all created Pipelines; Workflow objects are deleted after the specified amount of time.  The default value for this property is 24 hours. However, this property only affects the Pipeline API Server; the Argo Workflow controller does not use this property.

Author: ```suggestion
Kaptain has a global configuration property that allows you to set the default TTL for all created Pipelines; Workflow objects are deleted after the specified amount of time. The default value for this property is 24 hours. However, this property only affects the Pipeline API Server; the Argo Workflow controller does not use this property.
```

> While Notebook users can set this property via the DSL, they cannot specify a longer interval than the global property setting.  The Pipeline component always uses the smaller of the two specified TTL values between the DSL config and the global property.  Because workflow objects can be useful in debugging, we recommend choosing a conservative value for the global property value.

Author: ```suggestion
While Notebook users can set this property via the DSL, they cannot specify a longer interval than the global property setting. The Pipeline component always uses the smaller of the two specified TTL values between the DSL config and the global property. Because workflow objects can be useful in debugging, we recommend choosing a conservative value for the global property value.
```

<p class="message--note"><strong>NOTE: </strong> The metadata information that belongs to the Pipeline run is available in the Pipeline UI after the Argo Workflow custom resource is deleted. However, the Pipeline task (step)-level Pod information and logs are not available after the workflow deletion because these are retrieved directly from the workflow custom resource.</p>

To set the default TTL for all Pipelines, install or update Kaptain instance with the following parameter:
```bash
kubectl kudo update --instance kaptain --namespace kubeflow -p workflowsTTLSecondsAfterFinish="<ttl seconds>"
```

> ## Automatic cleanup for resources created by KFServing

Author: KFServing is correct, its a name..

### Overview

> KFServing serves models over HTTP(s) using the Knative Serving component. When a model is deployed to serving,

Author: KFServing is correct, its a name..

KFServing creates a set of Knative resources such as `Service`,`Route`, and `Revision`.

> There is always one Knative Service per model deployment, however, the number of Revisions can grow with time

Reviewer: ```suggestion
One Knative Service is always available per model deployment, however, the number of Revisions can grow with time
```

Over time, the number of Revisions and associated deployments can grow significantly; to avoid the undesirable overhead it
is recommended to garbage collect the outdated Revisions.

For example:
```
$> kubectl get revisions
NAME                          CONFIG NAME              K8S SERVICE NAME               GENERATION   READY
dev-mnist-predictor-c5kzr   dev-mnist-predictor   dev-mnist-predictor-c5kzr                1       True
dev-mnist-predictor-d6tdr   dev-mnist-predictor   dev-mnist-predictor-d6tdr                2       True
dev-mnist-predictor-tqzqw   dev-mnist-predictor   dev-mnist-predictor-tqzqw                3       True

$> kubectl get deployments
NAME                                                  READY   UP-TO-DATE   AVAILABLE   AGE
dev-mnist-predictor-c5kzr-deployment                   0/0        0            0       33m
dev-mnist-predictor-d6tdr-deployment                   0/0        0            0       18m
dev-mnist-predictor-tqzqw-deployment                   1/1        1            1       5m53s
```

### Configure Knative addon cleanup

> KFServing itself doesn’t provide controls for garbage collection of stale Revisions, however, the underlying

Author: KFServing is correct, its a name..

| retainSinceLastActiveTime       | 15h     | Duration since a revision was _active_ before considering it for the cleanup. An _active_ revision is the one that has the service network traffic routed to it. |
| maxNonActiveRevisions           | 1000    | Maximum number of non-active revisions to retain. If the maximum number of revisions reached, the oldest non-active revision will be deleted disregarding the other settings. |

### Update Knative addon configuration
To specify or update the Knative addon configuration, edit the `cluster.yaml` section and specify the values for
the garbage collection settings:
```
- configRepository: https://github.com/mesosphere/kubeaddons-kaptain
  configVersion: stable-1.20-1.3.0
  addonsList:
    - name: knative
      enabled: true
      values: |
        serving:
          gc:
            retainSinceCreateTime: "48h"
            retainSinceLastActiveTime: "15h"
            minNonActiveRevisions: "20"
            maxNonActiveRevisions: "1000"
```

After updating the settings, run `konvoy deploy addons` to apply the changes.

> <p class="message--note"><strong>NOTE: </strong>The Knative Controller automatically reloads configuration changes when the addon is updated. However, this doesn't trigger an automatic revision cleanup. The cleanup will be performed upon the next service version deployment.</p>

Author: ```suggestion
<p class="message--note"><strong>NOTE: </strong>The Knative Controller automatically reloads configuration changes when the addon is updated. However, this does not trigger an automatic revision cleanup. The cleanup will be performed upon the next service version deployment.</p>
```
